src/hooks/useTriggerTransition.js

```diff
@@ -0,0 +1,16 @@
+import { useContext } from "react";
+import { Context } from "../context/createTransitionContext";
+import { triggerTransition } from "../utils/triggerTransition";
+
+const useTriggerTransition = () => {
```

Comment:
I think the API for this should be that the options are passed in when the effect is used.

```js
const triggerTransition = useTriggerTransition({
  to: "/programmatic/b",
  exit: {
    length: 1,
    trigger: ({ exit, node }) => fade({ exit, node, direction: "out" })
  },
  entry: {
    length: 0.5,
    delay: 0.5,
    trigger: ({ exit, node }) => fade({ exit, node, direction: "in" })
  }
});

const progrToB = event => triggerTransition({ event });
```

We could then merge any additional options passed into the function that's returned

This is untested code but something like:

```js
const useTriggerTransition = initialOptions => {
  const context = useContext(Context);

  const programaticallyTriggerTransition = calledOptions => {
    triggerTransition({
      ...context,
      ...initialOptions,
      ...calledOptions
    });
  };

  return programaticallyTriggerTransition;
};
```

Reply:
Might be worth allowing to just pass an event in instead of options, to make usage simpler:

```js
const useTriggerTransition = initialOptions => {
  const context = useContext(Context);

  const programaticallyTriggerTransition = calledOptions => {
    // allow passing an event directly instead of options
    if (calledOptions instanceof Event) {
      calledOptions = {
        event: calledOptions
      };
    }

    triggerTransition({
      ...context,
      ...initialOptions,
      ...calledOptions
    });
  };

  return programaticallyTriggerTransition;
};
```

That way with this transition

```js
const triggerTransition = useTriggerTransition({
  to: "/programmatic/b",
  exit: {
    length: 1,
    trigger: ({ exit, node }) => fade({ exit, node, direction: "out" })
  },
  entry: {
    length: 0.5,
    delay: 0.5,
    trigger: ({ exit, node }) => fade({ exit, node, direction: "in" })
  }
});
```

You could either do

```js
<El onClick={event => triggerTransition({ event })} />
```

or

```js
<El onClick={triggerTransition} />
```

Reply:
Ah, just thought of one more thing! 😜 

There's an accessibility check in `gatsby-plugin-transition-link/src/components/TransitionLink.js` that checks for the OS setting `prefers-reduced-motion`, we should also respect that here in this hook, so if the setting is enabled, no animation will run, but navigation will still happen.

Reply:
Sorry for the delay :) I've implemented your suggestions, it all makes sense to me, especially the `onClick={triggerTransition}` shortcut.

I've also added the `prefers-reduced-motion` check, but I have some doubts. It makes perfect sense on the TransitionLink component, since the transition is triggered by a click. But if the transition is triggered programmatically there would probably be no click event (see `gatsby-plugin-transition-link/example/src/pages/programmatic/b.js`). If I return early (like in TransitionLink), nothing at all will happen.

What do you think of, instead, forcing the `exit` and `entry` properties to have a **0** duration? Something like:

```js
const useTriggerTransition = defaultOptions => {
  const context = useContext(Context);
  const programmaticallyTriggerTransition = calledOptions => {
    // If the user has set their browser or OS to prefers-reduced-motion
    // we should respect that.
    const forceZeroMotion = {};
    if (context.disableAnimation) {
      forceZeroMotion.exit = {
        length: 0,
        delay: 0
      };
      forceZeroMotion.entry = {
        length: 0,
        delay: 0
      };
    }

    // allow passing an event directly instead of options
    if (
      calledOptions instanceof Event ||
      (calledOptions.nativeEvent && calledOptions.nativeEvent instanceof Event)
    ) {
      calledOptions = {
        event: calledOptions
      };
    }

    triggerTransition({
      ...context,
      ...defaultOptions,
      ...calledOptions,
      ...forceZeroMotion
    });
  };
  return programmaticallyTriggerTransition;
};
```

My only doubt is that if the developer is unaware of this behaviour it could lead to some head scratching. Maybe a clear reference in the docs could help?

Reply:
I think it's best to stick with not running animations at all, as transitions can also trigger other functions that don't use the length and delay.

This is what I came up with:

```jsx
import { useContext } from 'react'
import { Context } from '../context/createTransitionContext'
import { triggerTransition } from '../utils/triggerTransition'
import { navigate } from 'gatsby'

const useTriggerTransition = defaultOptions => {
	const context = useContext(Context)

	const programmaticallyTriggerTransition = calledOptions => {
		// allow passing an event directly instead of options
		if (
			calledOptions instanceof Event ||
			(calledOptions.nativeEvent &&
				calledOptions.nativeEvent instanceof Event)
		) {
			calledOptions = {
				event: calledOptions,
			}
		}

		const options = {
			...defaultOptions,
			...calledOptions,
		}

		if (context.disableAnimation) {
			// If the user has set their browser or OS to prefers-reduced-motion
			// we should respect that.
			return navigate(options.to)
		}

		triggerTransition({
			...context,
			...options,
		})
	}

	return programmaticallyTriggerTransition
}

export { useTriggerTransition }

```

Seems to work! Let me know what you think.

Reply:
Also, thanks for making those changes! This is looking really great!

Reply:
Oh I hadn't thought of `navigate`, nice find! Will integrate tomorrow and update the PR 👌

Reply:
I've added 

```js
if (options.event) {
  options.event.persist();
  options.event.preventDefault();
}
```

before `navigate()` to catch some flickering if the page "gatsby-navigates" on click, but the click is not caught and it also "browser-navigates". For the rest, it seems to be working now!

Reply:
This is so close!! just left one more suggestion. The slight flickering is actually due to react transition group. The only way to prevent it is to put a background colour on the page unfortunately. In the future I may remove react-transition-group and add something custom in place, but for now it's unavoidable.